### PR TITLE
client, cmd, daemon, docs, integtration-tests, snap, store: present prices from store

### PR DESCRIPTION
**Pull request by pete-woods**
Display price information in find command. Add new price property to snap interface.

### CODE REVIEW

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by chipaca** [comment]
add to whitelist

File: snappy/snap_remote_repo.go [hunk 1]
```diff
@@ -212,7 +214,18 @@ func (s *SnapUbuntuStoreRepository) Snap(name, channel string) (*RemoteSnap, err
...
 	}
 
-	return NewRemoteSnap(detailsData), nil
+	return NewRemoteSnap(detailsData, suggestedCurrency), nil
+}
+
+func getSuggestedCurrency(h *http.Header) string {
+	s := h.Get("X-Suggested-Currency")
+	if s == "" {
+		s = os.Getenv("U1_SEARCH_CURRENCY")
```
Comment (kyrofa) [inline snappy/snap_remote_repo.go]: Consider adding a test for this case.

Comment (chipaca) [inline snappy/snap_remote_repo.go]: two things bother me here. One is that the store overrides the environment and not vice versa, which sounds strange to me; in general the local environment wins. What's the reasoning behind this?

The second is the `U1_` prefix, which is not related to anything in this code. We've been using `SNAPPY_` to namespace our environment variables, with the exception of the (legacy, and I hope we can change it sometime) `UBUNTU_STORE_ID`.

Comment (pete-woods) [inline snappy/snap_remote_repo.go]: I actually just copied this behaviour wholesale from the click scope without thinking about it too much. I'm tempted to just remove the environment variable from it altogether, unless we have a compelling use-case for it.

File: snappy/info.go [hunk 2]
```diff
@@ -70,7 +70,12 @@ func fullNameWithChannel(p *snap.Info) string {
 
 // ActiveSnapsByType returns all installed snaps with the given type
 func ActiveSnapsByType(snapTs ...snap.Type) (res []*Snap, err error) {
-	installed, err := NewLocalSnapRepository().Installed()
+	repo := NewLocalSnapRepository()
+	if repo == nil {
+		return nil, fmt.Errorf("No local snap repository")
+	}
+
+	installed, err := repo.Installed()
```
Comment (stevenwilkin) [inline snappy/info.go]: This refactor has nothing to do with the subject of the PR. Just saying ;)

Comment (pete-woods) [inline snappy/info.go]: This change was mistakenly left in. I will remove it.

File: cmd/snap/cmd_find.go [hunk 3]
```diff
@@ -75,11 +75,17 @@ func (x *cmdFind) Execute([]string) error {
...
+	fmt.Fprintln(w, i18n.G("Name\tVersion\tPrice\tSummary"))
 
 	for _, name := range names {
 		snap := snaps[name]
-		fmt.Fprintf(w, "%s\t%s\t%s\n", name, snap.Version, snap.Description)
+		var price string
+		if snap.Price > 0 {
+			price = fmt.Sprintf("%v", snap.Price)
+		} else {
+			price = "free"
```
Comment (zyga) [inline cmd/snap/cmd_find.go]: ``i18n.G("free")`` (or "gratis", whatever the design is)

File: snappy/snap_remote.go [hunk 4]
```diff
@@ -34,7 +34,8 @@ import (
 // RemoteSnap represents a snap available on the server
 type RemoteSnap struct {
 	// FIXME: exported for testing
-	Pkg remote.Snap
+	Pkg             remote.Snap
+	currentCurrency string
```
Comment (zyga) [inline snappy/snap_remote.go]: Hmm, this makes me feel uneasy. In what situation would we have multiple currencies in the system?

File: snappy/snap_remote_repo.go [hunk 5]
```diff
@@ -49,8 +49,8 @@ const (
 
 // NewRemoteSnap returns a new RemoteSnap from the given
 // remote.Snap data
-func NewRemoteSnap(data remote.Snap) *RemoteSnap {
-	return &RemoteSnap{Pkg: data}
+func NewRemoteSnap(data remote.Snap, suggestedCurrency string) *RemoteSnap {
```
Comment (zyga) [inline snappy/snap_remote_repo.go]: Currency should be a distinct type ``type Currency string``.

**Comment by fgimenez** [comment]
Hey @pete-woods , there's [an integration test failing](http://162.213.35.179:8080/job/github-snappy-integration-tests-cloud/735/consoleFull) (VPN required, you can search for `FAIL:` in the log to find the actual error) [this patch should fix it](http://paste.ubuntu.com/15464578/).

Don't hesitate to ping me if you need any help finding the error in the log or executing the tests locally. Thanks! :blush:

**Comment by chipaca** [comment]
@pete-woods needs conflicts resolving

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by pete-woods** [comment]
Seems like MO is back :)

**Comment by mvo5** [comment]
Code looks good from a first look, I wonder however if we need to display the suggestedCurrency (which I presume is server side selected based on geoip) next to the actual price value to avoid confusion with our users about the currency (and about cases where geoip might be incorrect)?

**Comment by pete-woods** [comment]
I was thinking that we could refine it in future to actually have snappy let user select which currency they want to use or a setting, or similar. So was really just trying to keep it as simple as possible until we had that part figured out.

File: po/snappy.pot [hunk 6]
```diff
@@ -619,6 +619,9 @@ msgstr  ""
 msgid   "data-size: %s\n"
 msgstr  ""
 
+msgid   "free"
```
Comment (zyga) [inline po/snappy.pot]: Add a translator hint that this is free as in beer, not as in freedom. It has different translations in many languages.

File: store/snap_remote_repo.go [hunk 7]
```diff
@@ -278,7 +293,17 @@ func (s *SnapUbuntuStoreRepository) Snap(name, channel string) (*snap.Info, erro
...
-	return infoFromRemote(detailsData), nil
+	suggestedCurrency := getSuggestedCurrency(&resp.Header)
+
+	return infoFromRemote(detailsData, suggestedCurrency), nil
+}
+
+func getSuggestedCurrency(h *http.Header) string {
+	s := h.Get("X-Suggested-Currency")
+	if s == "" {
+		s = "USD"
```
Comment (zyga) [inline store/snap_remote_repo.go]: Is this something we should do?

**Comment by zyga** [comment]
One thing wrt translations that bit me in the past. Otherwise +1

File: po/snappy.pot (hunk 6 above)
Comment (pete-woods) [inline po/snappy.pot]: Thanks for the hint! :)

**Comment by chipaca** [comment]
could you deconflict?

**Comment by pete-woods** [comment]
I've de-conflicted, and updated the REST API to pass through the prices map all the way through, so price presentation can be handled on the client side.

File: store/snap_remote_repo.go (hunk 7 above)
Comment (pete-woods) [inline store/snap_remote_repo.go]: It's what the current store-front on the phone does (unity-scope-click). I don't know what other behaviour to apply.

**Comment by niemeyer** [comment]
I'm working with Pete to tweak this and get it merged in smaller pieces.

**Pull request closed without merging**